// @electron-forge/maker-wix/dist/MakerWix.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _makerBase = _interopRequireDefault(require("@electron-forge/maker-base"));

require("colors");

var _logSymbols = _interopRequireDefault(require("log-symbols"));

var _path = _interopRequireDefault(require("path"));

var _creator = require("electron-wix-msi/lib/creator");

var _authorName = _interopRequireDefault(require("./util/author-name"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class MakerWix extends _makerBase.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "name", 'wix');

    _defineProperty(this, "defaultPlatforms", ['win32']);
  }

  isSupportedOnCurrentPlatform() {
    return process.platform === 'win32';
  }

  async make({
    dir,
    makeDir,
    targetArch,
    packageJSON,
    appName
  }) {
    const outPath = _path.default.resolve(makeDir, `wix/${targetArch}`);

    await this.ensureDirectory(outPath);
    let {
      version
    } = packageJSON;

    if (version.includes('-')) {
      // eslint-disable-next-line no-console
      console.warn(_logSymbols.default.warning, 'WARNING: WiX distributables do not handle prerelease information in the app version, removing it from the MSI'.yellow);
      version = this.normalizeWindowsVersion(version);
    }

    const creator = new _creator.MSICreator({
      description: packageJSON.description,
      name: appName,
      version,
      manufacturer: (0, _authorName.default)(packageJSON.author),
      exe: `${appName}.exe`,
      ...this.config,
      appDirectory: dir,
      outputDirectory: outPath
    });

    if (this.config.beforeCreate) {
      await Promise.resolve(this.config.beforeCreate(creator));
    }

    await creator.create();
    const {
      msiFile
    } = await creator.compile();
    return [msiFile];
  }

}

exports.default = MakerWix;
