// electron-wix-msi/lib/app.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const index_1 = require("./index");
const appDirectory = path.join(__dirname, '../../debug/sample-app');
const appIconPath = path.join(__dirname, '../../debug/sample-app/HelloWix.ico');
const o = {
    appDirectory,
    exe: `HelloWix.exe`,
    description: 'A sample app to debug wix',
    manufacturer: 'Electron Wix Corp',
    name: 'hellowix',
    outputDirectory: 'C:\\temp\\wix',
    programFilesFolderName: undefined,
    upgradeCode: undefined,
    version: '1.0.0-alpha',
    appIconPath
};
function build() {
    return __awaiter(this, void 0, void 0, function* () {
        const wix = new index_1.MSICreator(o);
        yield wix.create();
        yield wix.compile();
    });
}
build();
